src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1117,245 @@ private static string RegularExpressionFromFileSpec
-             *     *. to match foo
-             * 
-             * but 'foo' doesn't have a trailing '.' so we need to handle this while still being careful 
-             * not to match 'foo.txt'
-             */
-            if (filenamePart.EndsWith(".", StringComparison.Ordinal))
+        private static bool HasMisplacedRecursiveOperator(string str)
+        {
+            for (int i = 0; i < str.Length - 1; i++)
             {
-                filenamePart = filenamePart.Replace("*", "<:anythingbutdot:>");
-                filenamePart = filenamePart.Replace("?", "<:anysinglecharacterbutdot:>");
-                filenamePart = filenamePart.Substring(0, filenamePart.Length - 1);
-            }
+                bool isRecursiveOperator = str[i] == '*' && str[i + 1] == '*';
 
-            /*
-             * Now, build up the starting filespec but put tags in to identify where the fixedDirectory,
-             * wildcardDirectory and filenamePart are. Also tag the beginning of the line and the end of
-             * the line, so that we can identify patterns by whether they're on one end or the other.
-             */
-            StringBuilder matchFileExpression = new StringBuilder();
-            matchFileExpression.Append("<:bol:>");
-            matchFileExpression.Append("<:fixeddir:>").Append(fixedDirectoryPart).Append("<:endfixeddir:>");
-            matchFileExpression.Append("<:wildcarddir:>").Append(wildcardDirectoryPart).Append("<:endwildcarddir:>");
-            matchFileExpression.Append("<:filename:>").Append(filenamePart).Append("<:endfilename:>");
-            matchFileExpression.Append("<:eol:>");
+                // Check boundaries for cases such as **\foo\ and *.cs**
+                bool isSurroundedBySlashes = (i == 0 || FileUtilities.IsAnySlash(str[i - 1]))
+                                             && i < str.Length - 2 && FileUtilities.IsAnySlash(str[i + 2]);
 
-            /*
-             *  Call out our special matching characters.
-             */
-            foreach (var separator in directorySeparatorStrings)
-            {
-                matchFileExpression.Replace(separator, "<:dirseparator:>");
+                if (isRecursiveOperator && !isSurroundedBySlashes)
+                {
+                    return true;
+                }
             }
+            return false;
+        }
+
+        private static void AppendRegularExpressionFromFixedDirectory(StringBuilder regex, string fixedDir)
+        {
+            regex.Append(FileSpecRegexParts.BeginningOfLine);
 
             /*
              * Capture the leading \\ in UNC paths, so that the doubled slash isn't
              * reduced in a later step.
              */
-            matchFileExpression.Replace("<:fixeddir:><:dirseparator:><:dirseparator:>", "<:fixeddir:><:uncslashslash:>");
+            bool isUncPath = NativeMethodsShared.IsWindows && fixedDir.Length > 1
+                             && fixedDir[0] == '\\' && fixedDir[1] == '\\';
```

Comment:
Why not use Microsoft.Build.Shared.FileUtilities.IsUNCPath

Reply:
That one has different behavior. It returns false if any other ```\``` is in the string which seems like the wrong behavior for this looking at ```FileMatcher_Tests```, and also seems like the wrong behavior for determining if something is a UNC path but it's been part of ```NormalizePath()``` forever.